### PR TITLE
Verify more constants are not loaded at startup

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Loading different constants can be tricky with all the requires in Homebrew so we want to strengthen the existing checks to make sure that things are not getting required when there is a performance penalty. This expands the existing check to include more constants beyond `Formula` that we don't expect to be defined and that pull in a lot of other dependencies.

Note: This currently fails the `Tap` and `Homebrew::API` checks but those should resolve themselves after https://github.com/Homebrew/brew/pull/18010 gets merged in.
Note 2: This PR conflicts with https://github.com/Homebrew/brew/pull/18008 so we should wait until that one gets merged in.

### CODE REVIEW

File: Library/Homebrew/test/support/helper/cmd/brew-verify-undefined.rb
Comment (dduugg) [inline Library/Homebrew/test/support/helper/cmd/brew-verify-undefined.rb:14]: Note that https://github.com/Homebrew/brew/pull/18008 will require passing an `AbstractCommand` class when instantiating a `Parser`. (I'd like Parser to have a class-level `parse` method, but that's a massive refactor…)

Comment (apainintheneck) [reply Library/Homebrew/test/support/helper/cmd/brew-verify-undefined.rb:14]: @dduugg The plan is to wait until your PR gets merged in before rebasing this one. I didn't realize there a conflict before I opened it and there's no rush here. My hope is that the main logic shouldn't have to change too much.

**Review by MikeMcQuaid — APPROVED** [review]
Great idea, thanks for this @apainintheneck!

File: Library/Homebrew/test/support/helper/cmd/brew-verify-undefined.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/test/support/helper/cmd/brew-verify-undefined.rb:11]: Nice work @apainintheneck! Any other big constants you think we could add in here? I'm happy to take a look and suggest some if that's be helpful. My thinking would just be to add various `require` to `global.rb` and use `brew prof` to see which ones have the biggest impact on a `brew ruby -e 'puts :foo'` (or similar quick command that goes through Ruby-land) output.

Comment (apainintheneck) [reply Library/Homebrew/test/support/helper/cmd/brew-verify-undefined.rb:11]: There are a few more that we don't want to be loaded at start up but they are required by some of the existing files in this list so it's not strictly necessary to add them here. I'm thinking of `FormulaInstaller` and `Keg` in this case. That being said the check here is really simple so there is essentially no cost to us expanding this list. Feel free to take a look and add any that seem mildly suspicious to the list. `Homebrew::API` is an example of a borderline one that I decided to add on a whim.

Comment (MikeMcQuaid) [reply Library/Homebrew/test/support/helper/cmd/brew-verify-undefined.rb:11]: Thanks @apainintheneck. Inspired by your great work here: I've opened https://github.com/Homebrew/brew/pull/18065 as a follow-up.

I feel pretty confident after this that I can stop asking people to not add them to `global.rb` as CI will now catch this instead. Couldn't have done it without your idea here, wonderful stuff 👏🏻 😍

**Pull request merged**
